### ISSUE TITLE
Re-architecture of job workflow domain

### ISSUE THREAD

**Issue by aaraney** [NOAA-OWP/DMOD#329]
The primary goal of this issue is to describe our current job workflow architecture and propose moving to an event orchestration architecture. This issue is somewhere at the mid-to-high level to get the conversation started and get on the same page. The desired outcome of this issue is a fruitful discussion that leads us towards writing up one or more low level issues that describe tasks to accomplish this work and a specification for event broker messages and first system workflows.

But why? This sounds like a lot of work... and it is. So, i'll start by describing our current system and its pros and cons.

### Current System

Our current workflow procedures use choreography to accomplish a workflow task and is conceptually a global dictionary and global set that services spin-lock. After some time, a service will take the lock and grab all active jobs and check each job's phase. Work will be done if the service cares about that phase. For example, check that the data store has the required data to fulfill the job request. The service will then change the job's phase to the next phase in the state machine. In the previous example, the service might change the job phase from `AWAITING_DATA_CHECK` to `AWAITING_ALLOCATION` or to `DATA_UNPROVIDEABLE` if there was a failure.  This works and for small workflows is a good choice. It's simple and easy to follow especially if you have a state machine diagram that links back to the code.

Speaking of such a diagram, here is what ill call a _happy path_ state diagram. The diagram ignores failures and shows a job's state changes and the services that transition the state. The state is stored in Redis (our global dictionary) where we use the job's UUID as the key and a stringified version of the job object as the value. The current state of the job is held in the job object's `status` field (e.g. `MODEL_EXEC:AWAITING_DATA_CHECK`).

```mermaid
stateDiagram-v2
    state "JobPhase(JobExecPhase.INIT, JobExecStep.DEFAULT)" as s1
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_DATA_CHECK)" as s2
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_ALLOCATION)" as s3_1
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_PARTITIONING)" as s3_2

    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_DATA)" as s4
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_SCHEDULING)" as s5
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.SCHEDULED)" as s6
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.RUNNING)" as s7
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.COMPLETE)" as s8
    state if_state <<choice>>


    [*] --> s1: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L420">Created by SchedulerService</a>
    s1 --> s2: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L341">Transitioned by SchedulerService</a>
    s2 --> if_state: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/dataservice/dmod/dataservice/service.py#L956">Transitioned by DataService</a>
    if_state --> s3_1 : if cpu_count <= 1
    if_state --> s3_2 : if cpu_count > 1
    s3_2 --> s3_1: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/partitionerservice/dmod/partitionerservice/service.py#L469">Transitioned by PartitioningService</a>
    s3_1 --> s4: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L368">Transitioned by SchedulerService</a>
    s4 --> s5: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/dataservice/dmod/dataservice/service.py#L991">Transitioned by DataService</a>
    s5 --> s6: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L590">Transitioned by SchedulerService</a>
    s6 --> s7: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/monitor/dmod/monitor/que_monitor.py#L192">Transitioned by MonitorService</a>
    s7 --> s8: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/monitor/dmod/monitor/que_monitor.py#L206">Transitioned by MonitorService</a>
    s8 --> [*]

```

<details>
<summary>Full state diagram</summary>

Bug in github rendering of this. You can copy and paste this into a blank issue to view it for now. Just dont include the `<details>` and `<summary> tags and it will work. 

```mermaid
stateDiagram-v2
    state "JobPhase(JobExecPhase.INIT, JobExecStep.DEFAULT)" as s1
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_DATA_CHECK)" as s2
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_ALLOCATION)" as s3_1
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_PARTITIONING)" as s3_2

    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_DATA)" as s4
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.AWAITING_SCHEDULING)" as s5
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.SCHEDULED)" as s6
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.RUNNING)" as s7
    state "JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.COMPLETE)" as s8
    state if_state <<choice>>
    state partition_result <<choice>>
    state data_result <<choice>>
    state scheduler_result <<choice>>
    data_unprovidable: JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.DATA_UNPROVIDEABLE)
    partition_failure: JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.PARTITIONING_FAILED)
    data_failure: JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.DATA_FAILURE)
    failed: JobPhase(JobExecPhase.MODEL_EXEC, JobExecStep.FAILED)


    [*] --> s1: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L420">Created by SchedulerService</a>

    s1 --> s2: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L341">Transitioned by SchedulerService</a>

    s2 --> if_state: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/dataservice/dmod/dataservice/service.py#L956">Transitioned by DataService</a>

    if_state --> s3_1 : if cpu_count <= 1
    if_state --> s3_2 : if cpu_count > 1
    if_state --> data_unprovidable: Failure
    data_unprovidable --> [*]

    s3_2 --> partition_result: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/partitionerservice/dmod/partitionerservice/service.py#L469">Transitioned by PartitioningService</a>
    partition_result --> s3_1
    partition_result --> partition_failure: Failure
    partition_failure --> [*]

    s3_1 --> data_result: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L368">Transitioned by SchedulerService</a>
    data_result --> s4 
    data_result --> data_failure: Failure
    data_failure --> [*]

    s4 --> s5: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/services/dataservice/dmod/dataservice/service.py#L991">Transitioned by DataService</a>
    s5 --> s6: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/scheduler/dmod/scheduler/job/job_manager.py#L590">Transitioned by SchedulerService</a>

    s6 --> scheduler_result: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/monitor/dmod/monitor/que_monitor.py#L192">Transitioned by MonitorService</a>
    scheduler_result --> s7
    scheduler_result --> failed: Failure
    failed  --> [*]

    s7 --> s8: <a href="/NOAA-OWP/DMOD/blob/54799ceb771663f046ba48f8670db473678e06ca/python/lib/monitor/dmod/monitor/que_monitor.py#L206">Transitioned by MonitorService</a>
    s8 --> [*]

```

</details>

As requirements and services change, our current system becomes difficult to conceptually scale and requires broader knowledge of the system. This is mainly because each service contains logic about a job's next phase and the programmer has to be aware of local logic in each service to reason about the system. For example, let's say we add a forcing service that goes and gets atmospheric forcing data and puts it in the data store. From the data service example in the previous paragraph, the data service might add logic to conditionally check if forcings are available and if they aren't, change the job phase to `AWAITING_FORCINGS`. If the forcing service can fulfill the request, it might change the job phase back to `AWAITING_DATA_CHECK`. This seems sensible, however what if we want to use the forcing service to just retrieve forcings? Now the service has to have logic for both use cases. Meaning the service itself has to be aware of the workflows it might be used in. As more workflows are added, there will be a tradeoff between adding logic to a service to support a new workflow _or_ squeezing the new workflows into an existing workflow context even when it might not make the most sense. Moreover, service additions will likely require logic changes to other services and that just feels wrong to me.

Likewise, our current system is not the most scalable from a horizontal scaling perspective. Its worth noting that this is a side-effect of the current spin-lock implementation and could be resolved by moving to something like Redis streams, but to illustrate my point, services that could fan-out work, e.g. the data service, are not designed to do so now. Likewise, lock contention is linearly related to the number of services and service replicas. At this point, we don't _need_ the ability to fan-out and I think it's totally fair criticism to categorize this as engineering for a problem we don't yet have. I brought up these specific cons of our current system to advocate for the use of Redis streams or something like it. More about that in a bit...

A related aside, its a little unclear how we might rollback the system given some failure state if we needed to. Think a distributed transaction. Im sure we could add this in to make services responsible for rolling back data created for example, but _how_ to propagate that rollback across service boundaries is unclear to me. Anyways, continuing...

The last con I will bring up is our lack of providence and ability to reason with how our system got to its current state. Thinking back to the first paragraph in this section where I mentioned our system uses a global dictionary and set, state changes mutate objects in our dictionary. This makes it nearly impossible to know how a job got into its current state. We also don't have a great way at the moment of knowing how long each phase took (maybe we can do this with logs?). Having this kind of information would be really insightful for debugging and reproducibility, scientific data providence tracking, and powering features like realtime job progress and job progress notifications among other things.

TL;DR

Our current system is basically a global dictionary and set that services spin-lock to get work and transition a job's state machine. Our current design required local service-level logical reasoning to understand global workflow state transitions, requires modifying service logic to add or modify workflow phases (i.e. adding a new service), and is difficult to determine how the system got to the current state.

### Event Orchestration Architecture

So then, what is the proposed alternative? Switching to an event orchestration architecture. The main differences are:

- Communicate over a message broker (i.e. Redis streams), you don't have to do this, but it has so
many perks.
- Workflow logic is handled in one place and controlled by an Orchestration Service.
  - Commands to do work are published to the message broker. Example command could be
    `{"command": "CheckData",  "job_id": "42", "data_requirements": ...}` plus other data.
  - Resulting events are consumed and used to fire new commands.
- Services consume domain commands (`CheckData`), perform work, and publish resulting events (e.g. `{"event": "DataChecked", "successful": true, "job_id": "42"}`).
- Current system state can be derived from the actions that have previously occurred (commonly called a projection).
- Workflows are _more_ straightforward to rollback if required (see [saga](https://www.redhat.com/architect/pros-and-cons-saga-architecture-pattern)).

For the more visually inclined (like myself), I like [Derek Comartin's](https://codeopinion.com/event-choreography-orchestration-sagas/) description and visualization of Orchestration. Its about halfway down the linked page.

Other nice things event orchestration gives us:

- Reasonability: A single place to reason about the possible workflow states.
- Indirect workflow services are easy to create as they can access message broker topics (i.e. `JobProgressService`).
- The systems providence state is somewhat captured in the timeline of previous messages. For more
  robust tracking an [event store](https://en.wikipedia.org/wiki/Event_store) could be added to the
  Orchestration Service or provided as its own service.
- Reduce service coupling and increase workflow addition and modification flexibility.

Drawbacks:

- Reliance on canonical message model. If this is not documented well, it becomes really difficult
  to understand the events and shape of the events published and subscribed too by services.
- Does not scale well for non-workflow like tasks. Meaning, this approach should not be applied to the rest of dmod. However, having a message broker in the dmod stack (which we already have) enables other services to take advantage of the events being created in this bounded context.

With the goal of making this more tangible, the below _happy path_ sequence diagram shows the sequence of command events and result events that model our current system. Take the event names with a grain of salt, the goal of this is to show the concept not the minutia.  Ive purposefully not included the message schemas that services would publish and subscribe. I think defining those schema are a great next step to move this work forward. Not to add too much technical detail, but I am envisioning that we use [Redis streams](https://redis.io/docs/data-types/streams/) as the message broker simply because we already have it in the stack and it fits this use case well.

Legend:

- Topic denote a particular channel or stream.
- Solid lines represent publishing a message to a topic.
- Dashed lines represent reading a message from a topic.

```mermaid
sequenceDiagram
    participant rs as Request Service
    participant jt as Job Topic
    participant os as Orchestration Service

    participant dt as Data Topic
    participant ds as Data Service

    participant pt as Partition Topic
    participant ps as Partition Service

    participant st as Scheduler Topic
    participant ss as Scheduler Service

    participant ms as Monitor Service

    autonumber
    rs -)   jt: Add Job
    jt  --) os: Add Job
    os -)   jt: Job Added
    jt  --) rs: Job Added

    os -)   dt: Check Data
    dt --)  ds: Check Data
    ds -)   jt: Data Checked
    jt  --) os: Data Checked

    os -)   pt: Partition Realization Config
    pt --)  ps: Partition Realization Config
    ps -)   jt: Partition Complete
    jt  --) os: Partition Complete

    os -)   st: Allocate Compute
    st --)  ss: Allocate Compute
    ss -)   jt: Compute Allocated
    jt  --) os: Compute Allocated

    os -)   dt: Request Data
    dt --)  ds: Request Data
    ds -)   jt: Data Ready
    jt  --) os: Data Ready

    os -)   st: Scheduler Job
    st --)  ss: Scheduler Job
    ss -)   jt: Job Scheduled
    jt  --) os: Job Scheduled

    ms -)   jt: Job Started
    os --)  jt: Job Started

    ms -)   jt: Job Complete
    os --)  jt: Job Complete
```
